backends/sycl-ref/ceed-sycl-ref-qfunctioncontext.sycl.cpp

```diff
@@ -17,80 +17,218 @@
+      break;
+  }
+
+  return CEED_ERROR_SUCCESS;
 }
 
 //------------------------------------------------------------------------------
 // Check if data of given type needs sync
 //------------------------------------------------------------------------------
 static inline int CeedQFunctionContextNeedSync_Sycl(const CeedQFunctionContext ctx, CeedMemType mem_type, bool *need_sync) {
-  Ceed ceed;
-  CeedCallBackend(CeedQFunctionContextGetCeed(ctx, &ceed));
-  return CeedError(ceed, CEED_ERROR_BACKEND, "Ceed SYCL function not implemented");
+  CeedQFunctionContext_Sycl *impl;
+  CeedCallBackend(CeedQFunctionContextGetBackendData(ctx, &impl));
+
+  bool has_valid_data = true;
+  CeedCallBackend(CeedQFunctionContextHasValidData(ctx, &has_valid_data));
+  switch (mem_type) {
+    case CEED_MEM_HOST:
+      *need_sync = has_valid_data && !impl->h_data;
+      break;
+    case CEED_MEM_DEVICE:
+      *need_sync = has_valid_data && !impl->d_data;
+      break;
+  }
+
+  return CEED_ERROR_SUCCESS;
 }
 
 //------------------------------------------------------------------------------
 // Set data from host
 //------------------------------------------------------------------------------
 static int CeedQFunctionContextSetDataHost_Sycl(const CeedQFunctionContext ctx, const CeedCopyMode copy_mode, void *data) {
-  return CeedError(NULL, CEED_ERROR_BACKEND, "Ceed SYCL function not implemented");
+  CeedQFunctionContext_Sycl *impl;
+  CeedCallBackend(CeedQFunctionContextGetBackendData(ctx, &impl));
+
+  CeedCallBackend(CeedFree(&impl->h_data_owned));
+  switch (copy_mode) {
+    case CEED_COPY_VALUES:
+      size_t ctxsize;
+      CeedCallBackend(CeedQFunctionContextGetContextSize(ctx, &ctxsize));
+      CeedCallBackend(CeedMallocArray(1, ctxsize, &impl->h_data_owned));
+      impl->h_data_borrowed = NULL;
+      impl->h_data          = impl->h_data_owned;
+      memcpy(impl->h_data, data, ctxsize);
+      break;
+    case CEED_OWN_POINTER:
+      impl->h_data_owned    = data;
+      impl->h_data_borrowed = NULL;
+      impl->h_data          = data;
+      break;
+    case CEED_USE_POINTER:
+      impl->h_data_borrowed = data;
+      impl->h_data          = data;
+      break;
+  }
+
+  return CEED_ERROR_SUCCESS;
 }
 
 //------------------------------------------------------------------------------
 // Set data from device
 //------------------------------------------------------------------------------
 static int CeedQFunctionContextSetDataDevice_Sycl(const CeedQFunctionContext ctx, const CeedCopyMode copy_mode, void *data) {
+  CeedQFunctionContext_Sycl *impl;
+  CeedCallBackend(CeedQFunctionContextGetBackendData(ctx, &impl));
   Ceed ceed;
   CeedCallBackend(CeedQFunctionContextGetCeed(ctx, &ceed));
-  return CeedError(ceed, CEED_ERROR_BACKEND, "Ceed SYCL function not implemented");
+  Ceed_Sycl *sycl_data;
+  CeedCallBackend(CeedGetData(ceed, &sycl_data));
+
+  CeedCallSycl(ceed, sycl::free(impl->d_data_owned, sycl_data->sycl_context));
```

Comment:
Before freeing device memory, we should ensure that any work requiring the memory has completed.
For now, this can be done [like in CeedVector](https://github.com/argonne-lcf/libCEED/blob/4931d201cdeae5154a23bf13576b4fe9d5375903/backends/sycl-ref/ceed-sycl-vector.sycl.cpp#L694). When studying performance later, we can review the impact of these synchronizations.